test/server/listener_manager_impl_test.cc

```diff
@@ -213,6 +213,63 @@ class ListenerManagerImplWithRealFiltersTest : public ListenerManagerImplTest {
     return result;
   }
 
+  void
+  expectCreateListenSocket(const envoy::api::v2::core::SocketOption::SocketState& expected_state,
+                           Network::Socket::Options::size_type expected_num_options) {
+    EXPECT_CALL(listener_factory_, createListenSocket(_, _, _, true))
+        .WillOnce(Invoke([this, expected_num_options, &expected_state](
+                             Network::Address::InstanceConstSharedPtr, Network::Address::SocketType,
+                             const Network::Socket::OptionsSharedPtr& options,
+                             bool) -> Network::SocketSharedPtr {
+          EXPECT_NE(options.get(), nullptr);
+          EXPECT_EQ(options->size(), expected_num_options);
+          EXPECT_TRUE(
+              Network::Socket::applyOptions(options, *listener_factory_.socket_, expected_state));
+          return listener_factory_.socket_;
+        }));
+  }
+
+  void expectSetsockopt(NiceMock<Api::MockOsSysCalls>& os_sys_calls, int expected_sockopt_level,
+                        int expected_sockopt_name, int expected_value, int expected_num_calls) {
+    EXPECT_CALL(os_sys_calls,
+                setsockopt_(_, expected_sockopt_level, expected_sockopt_name, _, sizeof(int)))
+        .Times(expected_num_calls)
+        .WillRepeatedly(
+            Invoke([expected_value](int, int, int, const void* optval, socklen_t) -> int {
+              EXPECT_EQ(expected_value, *static_cast<const int*>(optval));
+              return 0;
+            }));
+  }
+
+  void expectSetsockopt(NiceMock<Api::MockOsSysCalls>& os_sys_calls,
+                        const Network::SocketOptionName& expected_option, int expected_value,
+                        int expected_num_calls) {
+    expectSetsockopt(os_sys_calls, expected_option.value().first, expected_option.value().second,
+                     expected_value, expected_num_calls);
+  }
+
+  void expectThrowForInvalidSockopt(const envoy::api::v2::Listener& listener) {
+    EXPECT_THROW_WITH_MESSAGE(manager_->addOrUpdateListener(listener, "", true), EnvoyException,
+                              "MockListenerComponentFactory: Setting socket options failed");
+    EXPECT_EQ(0U, manager_->listeners().size());
+  }
+
+  void endToEndSockoptTest(const envoy::api::v2::Listener& listener,
```

Comment:
Nit: I would call this `testSocketOption`, e2e has a very specific meaning in Google and Envoy that this probably doesn't capture.